hikari/internal/protocol.py

```diff
@@ -0,0 +1,118 @@
+if typing.TYPE_CHECKING:
+    _T = typing.TypeVar("_T")
+
+_Protocol = NotImplemented
+IGNORED_ATTRS = typing.EXCLUDED_ATTRIBUTES + ["__qualname__", "__slots__"]
+
+
+def _check_if_ignored(name: str) -> bool:
+    return name.startswith("_abc_") or name in IGNORED_ATTRS
+
+
+def _no_init(cls) -> None:
+    if type(cls)._is_protocol:
+        raise TypeError("Protocols cannot be instantiated")
+
+
+class _FastProtocol(abc.ABCMeta):
+    @staticmethod
+    def __new__(
+        cls: typing.Type[_T],
+        cls_name: str,
+        bases: typing.Tuple[typing.Type[typing.Any], ...],
+        namespace: typing.Dict[str, typing.Any],
+    ) -> _T:
+        global _Protocol
+
+        if _Protocol is NotImplemented:
+            if cls_name != "Protocol":
+                raise TypeError("First instance of _FastProtocol must be Protocol")
+
+            namespace = {"__init__": _no_init, "_is_protocol": True, "_attributes_": (), **namespace}
+            # noinspection PyRedundantParentheses
+            return (_Protocol := super().__new__(cls, cls_name, bases, namespace))
+
+        if _Protocol not in bases:
+            namespace["_is_protocol"] = False
+
+        else:
+            attributes = {attr for attr in namespace if not _check_if_ignored(attr)}
+            attributes.update(
+                annot for annot in namespace.get("__annotations__", {}).keys() if not _check_if_ignored(annot)
+            )
+
+            for base in bases:
+                # Make sure its a valid Protocol and not a class that mocks the behaviour
+                if not (issubclass(base, typing.Generic) and base._is_protocol):
+                    raise TypeError(f"Protocols can only inherit from other protocols, got {base!r}")
+                attributes.update(base._attributes_)
+
+            namespace = {
+                "__init__": _no_init,
+                "_is_protocol": True,
+                "_attributes_": tuple(attributes),
+                **namespace,
+            }
+
+        return super().__new__(cls, cls_name, bases, namespace)
+
+    def __instancecheck__(cls: _T, other: typing.Any) -> bool:
+        if not cls._is_protocol:
+            return super().__instancecheck__(other)
+
+        for i in cls._attributes_:
+            if not hasattr(other, i):
+                return False
+
+        return True
+
+
+class Protocol(typing.Generic, metaclass=_FastProtocol):
+    """A faster implementation of `typing.Protocol`."""
+
+    __slots__: typing.Sequence[str] = ()
+
+    def __init_subclass__(cls, *args, **kwargs):
+        super().__init_subclass__(*args, **kwargs)
+
+        if not cls._is_protocol:
+            return
+
+        def _subclass_hook(other: type) -> bool:
+            # Since we are overwriting, we need to check this
+            if not isinstance(other, type):
+                raise TypeError("issubclass() arg 1 must be a class")
+
+            return any(hasattr(other, a) for a in cls._attributes_) or not cls._attributes_
```

Comment:
wouldn't this be all rather than any?

Reply:
No, cause i want it to return as soon as it find one that is the same

Reply:
That doesn't exactly sound like how protocol instance or subclass checking works, wouldn't they usually only pass if all the attributes are present?

Reply:
No, `issubclass` checks if one of the attrs or annotations is there, if so, its a subclass, if not, its not

As I wrote this i realized that if they are all not present, it checks the actual inheritance, so will add that

Reply:
Does it? Doesn't look like the behaviour i'm seeing and if that was the case it'd break a ton of stuff that was using it since having one attribute incommon doesn't mean it's compatible
```py
>>> import typing
>>> @typing.runtime_checkable
class FooProto(typing.Protocol):
	def foo():
		...
	def bar():
		...

		
>>> class Bar():
	def foo():
		...

		
>>> class Foo():
	def foo():
		...
	def bar():
		...

		
>>> issubclass(Foo, FooProto)
True
>>> issubclass(Bar, FooProto)
False
```
is the behaviour i saw